homeassistant/components/stream/worker.py

```diff
@@ -154,43 +152,42 @@ def check_flush_part(self, packet: av.Packet) -> None:
                 stream_id=self._stream_id,
                 init=self._memory_file.getvalue(),
             )
-            self._segment_last_write_pos = byte_position
+            self._memory_file_pos = self._memory_file.tell()
+            self._part_start_dts = self._segment_start_dts
             # Fetch the latest StreamOutputs, which may have changed since the
             # worker started.
             for stream_output in self._outputs_callback().values():
                 stream_output.put(self._segment)
         else:  # These are the ends of the part segments
-            self._segment.parts.append(
-                Part(
-                    duration=float(
-                        (packet.dts - self._part_start_dts) * packet.time_base
-                    ),
-                    has_keyframe=self._part_has_keyframe,
-                    data=self._memory_file.getbuffer()[
-                        self._segment_last_write_pos : byte_position
-                    ].tobytes(),
-                )
-            )
-            self._segment_last_write_pos = byte_position
-            self._part_start_dts = packet.dts
-            self._part_has_keyframe = False
+            self.flush_part(packet, last_part=False)
 
-    def flush(self, duration: Fraction, packet: av.Packet) -> None:
-        """Create a segment from the buffered packets and write to output."""
-        self._av_output.close()
+    def flush_part(self, packet: av.Packet, last_part: bool) -> None:
+        """Output a part from the most recent bytes in the memory_file."""
         assert self._segment
-        self._segment.duration = float(duration)
-        # Also flush the part segment (need to close the output above before this)
+        self._memory_file.seek(self._memory_file_pos)
```

Comment:
We have to seek because the file pointer is also getting updated by the writer ie pyav. So every time we give it a `seek` we are actually seeking back from the end to the `seek._memory_file_pos`, and the read automatically sticks it back at the end again for the writer to pick up where it left off later.

Reply:
Thanks for the explanation, that makes sense.